### PR TITLE
Defect/CPLP-1944-Email-Key-Value-Missing

### PR DESCRIPTION
**Pull request by qxz2mqe**
https://jira.catena-x.net/browse/CPLP-1944

### CODE REVIEW

File: src/marketplace/Offers.Library/Service/OfferService.cs [hunk 1]
```diff
@@ -667,8 +660,15 @@ private async Task SendMail(IDictionary<string,IEnumerable<string>> receiverUser
...
             .GetCompanyUserEmailForCompanyAndRoleId(roleData, companyId);
-        await foreach (var receiver in companyUserWithRoleIdForCompany)
+        await foreach (var (receiver,FirstName,LastName) in companyUserWithRoleIdForCompany)
         {
+            var mailParams = new Dictionary<string, string>
+            {
+                { "offerName", offerName },
+                { "url", basePortalAddress },
+                { "declineMessage", message },
+                { "offerProviderName", string.Concat(FirstName, LastName)??"Service Manager" },
```
Comment (ntruchsess) [inline src/marketplace/Offers.Library/Service/OfferService.cs:670]: string.Concat will do the same as operator +. There will be no separating space in between FirstName and LastName which will look rather odd. In addition to that the result of string.concat is never null as it uses String.Empty for any null value and would return string.empty if both arguments are null. Please add a private method similar to https://github.com/catenax-ng/tx-portal-backend/blob/a2b9277b48d403c69db9ec728043236bdfc92081/src/provisioning/Provisioning.Library/Service/UserProvisioningService.cs#L202 to build a suitable name-string

File: src/portalbackend/PortalBackend.DBAccess/Repositories/UserRepository.cs [hunk 2]
```diff
@@ -357,14 +357,14 @@ public void AttachAndModifyCompanyUser(Guid companyUserId, Action<CompanyUser> s
...
         _dbContext.CompanyUsers
-            .Where(x => 
+            .Where(x =>
                 x.CompanyId == companyId &&
-                x.CompanyUserStatusId == CompanyUserStatusId.ACTIVE && 
+                x.CompanyUserStatusId == CompanyUserStatusId.ACTIVE &&
                 x.UserRoles.Any(ur => userRoleIds.Contains(ur.Id)) &&
                 x.Email != null)
-            .Select(x => x.Email!)
+            .Select(x => new ValueTuple<string, string, string>(x.Email!, x.Firstname!, x.Lastname!))
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/UserRepository.cs:367]: the !-operators are incorrect, you have to declare the returntype with nullable elements like so:
```(string? Email, string? FirstName, string? LastName)```

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/Offers.Library/Service/OfferService.cs (hunk 1 above)
Comment (qxz2mqe) [reply src/marketplace/Offers.Library/Service/OfferService.cs]: fixed

File: src/portalbackend/PortalBackend.DBAccess/Repositories/UserRepository.cs (hunk 2 above)
Comment (qxz2mqe) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/UserRepository.cs]: fixed

**Comment by jjeroch** [comment]
2 PRs for the same ticket? 
https://github.com/catenax-ng/tx-portal-backend/pull/25

File: src/marketplace/Offers.Library/Service/OfferService.cs [hunk 3]
```diff
@@ -667,11 +661,33 @@ private async Task SendMail(IDictionary<string,IEnumerable<string>> receiverUser
...
-        await foreach (var receiver in companyUserWithRoleIdForCompany)
+        await foreach (var (receiver,FirstName,LastName) in companyUserWithRoleIdForCompany)
         {
-            await _mailingService.SendMails(receiver, mailParams, new List<string> { "offer-request-decline" }).ConfigureAwait(false);
+            var mailParams = new Dictionary<string, string>
+            {
+                { "offerName", offerName },
+                { "url", basePortalAddress },
+                { "declineMessage", message },
+                { "offerProviderName", CreateNameString(FirstName, LastName)},
```
Comment (Phil91) [inline src/marketplace/Offers.Library/Service/OfferService.cs:671]: `CreateNameString(FirstName, LastName)` can be shortend to something like: `var userName = string.Join(" ", new[] { receiver.Name, receiver.LastName }.Where(item => !string.IsNullOrWhiteSpace(item)));` than the assignement here would be `!string.IsNullOrWhiteSpace(userName) ? userName : "Service Manager"`

**Review by Phil91 — CHANGES_REQUESTED** [review]

File: src/marketplace/Offers.Library/Service/OfferService.cs (hunk 3 above)
Comment (ntruchsess) [reply src/marketplace/Offers.Library/Service/OfferService.cs:679]: I like this recommendation :-)

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**